capa/rules/__init__.py

```diff
@@ -1401,154 +1423,319 @@ def __getitem__(self, rulename):
+        """
+        Score the given feature by how "uncommon" it is, where a higher score is more uncommon.
+        Features that are not good for indexing will have a low score, or 0.
         """
-        split the given rules into two structures:
-          - "easy rules" are indexed by feature,
-            such that you can quickly find the rules that contain a given feature.
-          - "hard rules" are those that contain substring/regex/bytes features or match statements.
-            these continue to be ordered topologically.
 
-        a rule evaluator can use the "easy rule" index to restrict the
-        candidate rules that might match a given set of features.
+        #
+        # Today, these scores are manually assigned by intuition/experience/guesswork.
+        # We could do a large-scale feature collection and use the results to assign scores.
+        #
 
-        at this time, a rule evaluator can't do anything special with
-        the "hard rules". it must still do a full top-down match of each
-        rule, in topological order.
+        if isinstance(
+            node,
+            capa.features.common.MatchedRule,
+        ):
+            # If present, other rule must match before this one, in same scope.
+            # Use score from that rule, which will have already been processed due to topological sorting.
+            # Otherwise, use a default score of 5.
+            assert isinstance(node.value, str)
+            return scores_by_rule.get(node.value, 5)
+
+        elif isinstance(node, (capa.features.insn.Number, capa.features.insn.OperandNumber)):
+            v = node.value
+            assert isinstance(v, int)
+
+            if -0x8000 <= v <= 0x8000:
+                # Small numbers are probably pretty common, like structure offsets, etc.
+                return 3
+
+            if 0xFFFF_FF00 <= v <= 0xFFFF_FFFF:
+                # Numbers close to u32::max_int are also probably pretty common,
+                # like signed numbers close to 0 that are stored as unsigned ints.
+                return 3
+
+            if 0xFFFF_FFFF_FFFF_FF00 <= v <= 0xFFFF_FFFF_FFFF_FFFF:
+                # Like signed numbers closed to 0 that are stored as unsigned long ints.
+                return 3
+
+            # Other numbers are assumed to be uncommon.
+            return 7
+
+        elif isinstance(node, (capa.features.common.Substring, capa.features.common.Regex, capa.features.common.Bytes)):
+            # Scanning features (non-hashable), which we can't use for quick matching/filtering.
+            return 0
+
+        C = node.__class__
+        return {
+            #
+            # Very uncommon features that are probably very selective in capa's domain.
+            # When possible, we want rules to be indexed by these features.
+            #
+            capa.features.common.String: 9,
+            capa.features.insn.API: 8,
+            capa.features.file.Export: 7,
```

Comment:
String gets a high score because my intuition is that each string instance that we reference in a rule is very selective. That is, we don't expect strings to be shared asking rules very often. Therefore, when given the option, we should prefer to use an instance of a string to represent a rule, rather than a mnemonic, etc. 

Integer constants are probably reused occasionally.  APIs sometimes. Mnemonics all the time.